### PR TITLE
Pull request for Contiv Ansible code integration into Openshift Ansible

### PR DESCRIPTION
**Pull request by srampal**
Contiv ansible is disabled by default when Openshift SDN is used. To use contiv SDN, you need to set the following variables in the inventory file

openshift_use_openshift_sdn=False
openshift_use_contiv=True
os_sdn_network_plugin_name='cni'

There are also some additional contiv specific variables that can be configured once contiv SDN is enabled.

### CODE REVIEW

**Comment by DanyC97** [comment]
@srampal can the file `roles/contiv_facts/tasks/main.yml.bak ` be deleted? (duplicate of main.yml )

**Comment by DanyC97** [comment]
also @srampal check this [3rd party network guidelines](https://github.com/openshift/origin/pull/12981/files) (not merged yet though)

**Comment by jcpowermac** [comment]
@srampal can you please fix these below
```
roles/contiv/defaults/main.yaml:
  2:1       warning  missing document start "---"  (document-start)
  17:35     error    trailing spaces  (trailing-spaces)
  65:59     error    trailing spaces  (trailing-spaces)
roles/contiv/tasks/aci.yml:
  1:1       warning  missing document start "---"  (document-start)
roles/contiv/tasks/cleanup.yml:
  11:6      error    wrong indentation: expected 4 but found 5  (indentation)
  14:7      error    wrong indentation: expected 4 but found 6  (indentation)
  19:6      error    wrong indentation: expected 4 but found 5  (indentation)
  23:7      error    wrong indentation: expected 4 but found 6  (indentation)
  29:7      error    wrong indentation: expected 4 but found 6  (indentation)
  38:6      error    wrong indentation: expected 4 but found 5  (indentation)
  40:7      error    wrong indentation: expected 4 but found 6  (indentation)
  45:7      error    wrong indentation: expected 4 but found 6  (indentation)
  56:6      error    wrong indentation: expected 4 but found 5  (indentation)
  58:7      error    wrong indentation: expected 4 but found 6  (indentation)
roles/contiv/tasks/default_network.yml:
  18:1      error    too many blank lines (1 > 0)  (empty-lines)
roles/contiv/tasks/delete_contiv.yml:
  1:1       warning  missing document start "---"  (document-start)
  31:1      error    too many blank lines (1 > 0)  (empty-lines)
roles/contiv/tasks/download_bins.yml:
  1:1       warning  missing document start "---"  (document-start)
roles/contiv/tasks/main.yml:
  1:1       warning  missing document start "---"  (document-start)
roles/contiv/tasks/netmaster.yml:
  1:1       warning  missing document start "---"  (document-start)
roles/contiv/tasks/netmaster_iptables.yml:
  18:1      warning  comment not indented like content  (comments-indentation)
roles/contiv/tasks/netplugin.yml:
  1:1       warning  missing document start "---"  (document-start)
roles/contiv/tasks/netplugin_iptables.yml:
  20:1      warning  comment not indented like content  (comments-indentation)
roles/contiv/tasks/ovs.yml:
  1:1       warning  missing document start "---"  (document-start)
roles/contiv/tasks/pkgMgrInstallers/centos-install.yml:
  17:8      error    wrong indentation: expected 4 but found 7  (indentation)
  22:8      error    wrong indentation: expected 4 but found 7  (indentation)
  33:8      error    wrong indentation: expected 4 but found 7  (indentation)
roles/contiv_facts/tasks/fedora-install.yml:
  2:3       error    wrong indentation: expected 0 but found 2  (indentation)
```

File: roles/contiv/tasks/delete_contiv.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+- name: delete contiv
```
Comment (jcpowermac) [inline roles/contiv/tasks/delete_contiv.yml]: @srampal playbooks really shouldn't be in tasks directory.  Maybe move this to playbooks/adhoc

File: roles/contiv/tasks/cleanup.yml [hunk 2]
```diff
@@ -0,0 +1,58 @@
+---
```
Comment (jcpowermac) [inline roles/contiv/tasks/cleanup.yml]: @srampal playbooks really shouldn't be in tasks directory. Maybe move this to playbooks/adhoc

**Comment by jcpowermac** [comment]
@DanyC97 Contiv doesn't support TLS/SSL with etcd.  I also have been told that it is preferred not to use OpenShift/kube etcd instance.

**Comment by srampal** [comment]
@DanyC97 
1) Will make your suggested changes
2) Yes, we are aware of the networking requirements doc from  @rajatchopra  and have worked closely with @jcpowermac  and him to confirm we meet all the networking requirements. We have also demo'ed these to Joe and Rajat including some end to end app deployment scenarios over contiv.
3) On documentation, we have been told by RH to hold off until support for Openshift Enterprise becomes official and GA ... we are looking at additional options for documentation until then and will be happy to go over the setup of the inventory file variables etc with you offline. Feel free to join our contiv slack channel contiv.slack.com (contiv-users channel) or contact me directly until then.
3) On etcd, we use a separate etcd instance as Joe indicated.

**Comment by srampal** [comment]
@jcpowermac 

1) Making the changes you mentioned
2) Wrt the delete_contiv.yml file, I dont see any other vendor plugin cleanup files/ logic in playbooks/adhoc either, so for the time being, we could maybe retain this file under contiv/tasks ? Ideally we should merge this in with contiv/tasks/cleanup.yml in any case. Hence thinking that we retain delete_contiv.yml under contiv/tasks for the time being with this commit and later merge it in with contiv/tasks/cleanup.yml or other similar options ..

**Comment by srampal** [comment]
@sdodson please provide any thoughts you may have on where to place the cleanup/ uninstall code. In any case the playbooks under playbooks/adhoc are not officially supported and only Openshift Origin web site talks about using playbooks/adhoc/uninstall.yml in order to uninstall Openshift.

I will resubmit the PR with the changes suggested by @jcpowermac and @DanyC97  ... please provide any additional comments you may have.

Note: For now contiv is supported for Openshift Origin only, although it should work for Enterprise as well. Full official OSE support is planned a little later so we do have opportunity to make any additional adjustments/ enhancements until then if needed.

File: roles/contiv/tasks/cleanup.yml (hunk 2 above)
Comment (srampal) [inline roles/contiv/tasks/cleanup.yml]: Moved to playbooks/adhoc/contiv  in the latest commit as suggested ...

**Comment by srampal** [comment]
@DanyC97   and @jcpowermac   made the changes you suggested (except the vars file suggested by @DanyC97 (we can do that later once we consolidate all config parms for the different contiv networking modes). Until then we can still use the key contiv config variables directly in the top inventory file as we will describe in the documentation. Currently only the VLAN encap mode is supported and the documentation shall describe exactly the config parameters to use in the inventory file for this to work.

File: playbooks/adhoc/contiv/delete_contiv.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        - /opt/cni
+        - /opt/contiv
+        - /etc/systemd/system/netmaster.service
+        - /etc/systemd/system/netplugin.service
+        - /etc/systemd/system/contiv-etcd.service
+        - /etc/systemd/system/contiv-etcd.service.d
+        - /var/lib/contiv-etcd
+        - /etc/default/netmaster
+        - /etc/default/netplugin
+        - /etc/openvswitch/conf.db
```
Comment (DanyC97) [inline playbooks/adhoc/contiv/delete_contiv.yml]: @srampal small one but to be consistent with the other code, can you align all the with_items above like
```
with_items:
- /opt/cni
- /opt/contiv
- /etc/systemd/system/netmaster.service
- /etc/systemd/system/netplugin.service
....
```

File: roles/contiv/handlers/main.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+- name: reload systemd
+  command: systemctl --system daemon-reload
+
+- name: restart netmaster
+  service: name=netmaster state=restarted
+  when: netmaster_started.changed == false
+
+- name: restart netplugin
+  service: name=netplugin state=restarted
+  when: netplugin_started.changed == false
```
Comment (DanyC97) [inline roles/contiv/handlers/main.yml]: just cosmetic but will be nice to be consistent with the format - i prefer
```
- name: restart netplugin
  service: 
    name: netplugin 
    state: restarted
  when: netplugin_started.changed == false

```
but regardless just pick one and be consistent all the way through, currently is a mix ;)

File: roles/contiv/tasks/default_network.yml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+---
+# Is there a reason shell is used here instead of command?
+- name: Contiv | Wait for netmaster
+  shell: 'netctl --netmaster "http://{{ inventory_hostname }}:{{ netmaster_port}}" tenant ls'
+  register: tenant_result
```
Comment (DanyC97) [inline roles/contiv/tasks/default_network.yml]: be consistent and add a space like  `{{ netmaster_port }}`

File: roles/contiv/tasks/default_network.yml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  until: tenant_result.stdout.find("default") != -1
+  retries: 9
+  delay: 10
+
+- name: Contiv | Check if default-net exists
+  shell: 'netctl --netmaster "http://{{ inventory_hostname }}:{{ netmaster_port}}" net ls'
```
Comment (DanyC97) [inline roles/contiv/tasks/default_network.yml]: likewise be consistent as above

File: roles/contiv/tasks/default_network.yml [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+  register: net_result
+
+- name: Contiv | Create default-net
+  shell: 'netctl --netmaster "http://{{ inventory_hostname }}:{{ netmaster_port}}" net create --subnet={{contiv_default_subnet}} -e {{contiv_encap_mode}} -p {{contiv_default_network_tag}} --gateway={{contiv_default_gw}} default-net'
+  when: net_result.stdout.find("default-net") == -1
```
Comment (DanyC97) [inline roles/contiv/tasks/default_network.yml]: as above add space between `{{ X  }}`

File: roles/contiv/tasks/netmaster_iptables.yml [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+  service: name=iptables enabled=yes state=started
+
+- name: Netmaster IPtables | Open Netmaster with iptables
+  command: /sbin/iptables -I INPUT 1 -p tcp --dport {{ item }} -j ACCEPT -m comment --comment "contiv"
+  with_items:
+    - "{{ netmaster_port }}"
+    - "{{ contiv_rpc_port1 }}"
+    - "{{ contiv_rpc_port2 }}"
+    - "{{ contiv_rpc_port3 }}"
+  when: iptablesrules.stdout.find("contiv") == -1
```
Comment (DanyC97) [inline roles/contiv/tasks/netmaster_iptables.yml]: align under `with_items` as my comment above

File: roles/contiv/tasks/netplugin_iptables.yml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+- name: Netplugin IPtables | Open Netmaster with iptables
+  command: /sbin/iptables -I INPUT 1 -p tcp --dport {{ item }} -j ACCEPT -m comment --comment "contiv"
+  with_items:
+    - "{{ netmaster_port }}"
+    - "{{ contiv_rpc_port1 }}"
+    - "{{ contiv_rpc_port2 }}"
+    - "{{ contiv_rpc_port3 }}"
+    - "{{ contiv_etcd_port }}"
+    - "{{ kube_master_api_port }}"
+  when: iptablesrules.stdout.find("contiv") == -1
```
Comment (DanyC97) [inline roles/contiv/tasks/netplugin_iptables.yml]: align under `with_items`

File: roles/contiv/templates/contiv.cfg.j2 [hunk 10]
```diff
@@ -0,0 +1,6 @@
+{
+  "K8S_API_SERVER": "https://{{ hostvars[groups['masters'][0]]['ansible_' + netmaster_interface].ipv4.address }}:{{ kube_master_api_port }}",
+  "K8S_CA": "{{ openshift.common.config_base}}/node/ca.crt",
+  "K8S_KEY": "{{ openshift.common.config_base}}/node/system:node:{{ openshift.common.hostname }}.key",
+  "K8S_CERT": "{{ openshift.common.config_base}}/node/system:node:{{ openshift.common.hostname }}.crt"
```
Comment (DanyC97) [inline roles/contiv/templates/contiv.cfg.j2]: as above add space between `{{ X  }}`

File: roles/contiv/tasks/default_network.yml [hunk 11]
```diff
@@ -0,0 +1,16 @@
+---
+# Is there a reason shell is used here instead of command?
```
Comment (sdodson) [inline roles/contiv/tasks/default_network.yml]: No, I think command should work for all of the tasks here. Shell is only necessary when you expect shell variable evaluation or pipes from one command to another.

File: roles/contiv/tasks/ovs.yml [hunk 12]
```diff
@@ -0,0 +1,24 @@
+---
+- include: packageManagerInstall.yml
+  when: source_type == "packageManager"
+  tags:
+    - binary-update
+
+- name: OVS | Configure selinux for ovs
+  shell: semanage permissive -a openvswitch_t
```
Comment (sdodson) [inline roles/contiv/tasks/ovs.yml]: Why is this necessary? We run ovs with selinux enforcing.

File: roles/contiv/tasks/ovs.yml [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+
+- name: OVS | Start ovs
+  service: name=openvswitch state=started
+  register: ovs_started
+
+- name: OVS | Configure ovs
+  shell: "ovs-vsctl set-manager {{ item }}"
+  with_items:
+    - "tcp:127.0.0.1:6640"
+    - "ptcp:6640"
```
Comment (sdodson) [inline roles/contiv/tasks/ovs.yml]: Does this need to be persisted in some way or does ovs-vsctl persist it in a config file somewhere?

File: roles/contiv/tasks/pkgMgrInstallers/centos-install.yml [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+- name: PkgMgr CentOS | Install ovs
+  yum:
+    pkg=openvswitch
+    state=latest
+  environment:
+    http_proxy: "{{ http_proxy|default('') }}"
+    https_proxy: "{{ https_proxy|default('') }}"
+    no_proxy: "{{ no_proxy|default('') }}"
+  tags:
+    - ovs_install
```
Comment (sdodson) [inline roles/contiv/tasks/pkgMgrInstallers/centos-install.yml]: We include openvswitch in our channels or in channels that are enabled as pre-requisites to OpenShift installation. Would it be better to use that version? I'd prefer not grabbing this from an openstack repo.  OCP 3.5 will have OVS 2.6, all previous releases OVS 2.4.

File: roles/contiv/templates/contiv.cfg.j2 [hunk 15]
```diff
@@ -0,0 +1,6 @@
+{
+  "K8S_API_SERVER": "https://{{ hostvars[groups['masters'][0]]['ansible_' + netmaster_interface].ipv4.address }}:{{ kube_master_api_port }}",
```
Comment (sdodson) [inline roles/contiv/templates/contiv.cfg.j2]: This will just target the ip address of the first master, it may work in most situations but if that api server goes down and there are other api servers still online contiv will fail while the rest of the cluster stays up. Better to point at the cluster hostname. Can you try this?

```
"K8S_API_SERVER": "https://{{ hostvars[groups['masters'][0]].openshift.master.cluster_hostname }}:{{ kube_master_api_port }}",
```

File: roles/contiv/tasks/cleanup.yml (hunk 2 above)
Comment (sdodson) [inline roles/contiv/tasks/cleanup.yml]: Was this commit pushed? I don't see the change.

**Comment by sdodson** [comment]
Initial review complete, I'd like to run this on my own to verify, also +1 to all of the formatting/indentation changes that @DanyC97 suggested that seem unaddressed so far.

File: playbooks/adhoc/contiv/delete_contiv.yml (hunk 3 above)
Comment (srampal) [inline playbooks/adhoc/contiv/delete_contiv.yml]: Ok

File: roles/contiv/handlers/main.yml (hunk 4 above)
Comment (srampal) [inline roles/contiv/handlers/main.yml]: The convention used in all the contiv role files matches with "service: name=netplugin state=restarted".  Not clear when you say its a mix ...

File: roles/contiv/tasks/default_network.yml (hunk 5 above)
Comment (srampal) [inline roles/contiv/tasks/default_network.yml]: ok

File: roles/contiv/tasks/default_network.yml (hunk 6 above)
Comment (srampal) [inline roles/contiv/tasks/default_network.yml]: ok

File: roles/contiv/tasks/default_network.yml (hunk 7 above)
Comment (srampal) [inline roles/contiv/tasks/default_network.yml]: ok

File: roles/contiv/tasks/netmaster_iptables.yml (hunk 8 above)
Comment (srampal) [inline roles/contiv/tasks/netmaster_iptables.yml]: ok

File: roles/contiv/tasks/netplugin_iptables.yml (hunk 9 above)
Comment (srampal) [inline roles/contiv/tasks/netplugin_iptables.yml]: ok

File: roles/contiv/templates/contiv.cfg.j2 (hunk 10 above)
Comment (srampal) [inline roles/contiv/templates/contiv.cfg.j2]: ok

File: roles/contiv/tasks/netmaster.yml [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+
+- include: netmaster_iptables.yml
+  when: not has_firewalld and has_iptables
+
+- name: Netmaster | Check is /etc/hosts file exists
+  stat: path=/etc/hosts
+  register: hosts
+
+- name: Netmaster | Create hosts file if it is not present
+  file: path=/etc/hosts state=touch
```
Comment (DanyC97) [inline roles/contiv/tasks/netmaster.yml]: @srampal wrt to consistency compare this with ... see below

File: roles/contiv/tasks/netmaster.yml [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+    state=present
+  when: hostvars[item]['ansible_' + netmaster_interface].ipv4.address is defined
+  with_items: groups['masters']
+
+- name: Netmaster | Create netmaster symlinks
+  file:
+    src: "{{ contiv_current_release_directory }}/{{ item }}"
+    dest: "{{ bin_dir }}/{{ item }}"
+    state: link
+  with_items:
```
Comment (DanyC97) [inline roles/contiv/tasks/netmaster.yml]: @srampal wrt to consistency ....compare with above

File: roles/contiv/handlers/main.yml (hunk 4 above)
Comment (DanyC97) [inline roles/contiv/handlers/main.yml]: @srampal i went over and tagged the mix , hth. 

Also i'll point out that as general rule (at least that is how i personally take it) will be nice (*not mandatory*) for everyone (3rd party network providers etc ) to match their convention with the target repo/ project - in this case current one

Why would you want to do so: _only to play nice with OpenShift folks and help them have a tidy up project_ .

File: roles/contiv/tasks/ovs.yml (hunk 12 above)
Comment (srampal) [inline roles/contiv/tasks/ovs.yml]: At the moment this is necessary as long as we install ovs separate from Openshift. Without this the contiv data plane does not work correctly. When we switch to using the ovs that is already installed by openswitch (see the other comment on this), at that point we can decide whether to retain this or not. However for now, this is necessary for correct functional operation of Contiv.

File: roles/contiv/tasks/pkgMgrInstallers/centos-install.yml (hunk 14 above)
Comment (srampal) [inline roles/contiv/tasks/pkgMgrInstallers/centos-install.yml]: Agreed, we will move to the ovs installed by openshift in future. 

However currently we have done a lot of testing with ovs 2.5 which we install currently as part of contiv installation). This gives us confidence in the control plane + data plane combination that we have tested thoroughly and our QA automated testing suite also uses. So we would like to go ahead with our OVS for now and not hold up this commit/ merge. Once we have had sufficient experience testing with ovs 2.4 or 2.6 as deployed through openswitch, we will definitely switch to that version and not install our own ovs 2.5.

File: roles/contiv/tasks/default_network.yml (hunk 11 above)
Comment (srampal) [inline roles/contiv/tasks/default_network.yml]: ok, thanks.

File: roles/contiv/tasks/pkgMgrInstallers/centos-install.yml (hunk 14 above)
Comment (sdodson) [inline roles/contiv/tasks/pkgMgrInstallers/centos-install.yml]: @rajatchopra Do you have an opinion one way or another here? Do we let OVS from Centos OpenStack repos for now and switch to OVS from OpenShift repos in the future or do we need to do that now?

File: roles/contiv/templates/contiv.cfg.j2 (hunk 15 above)
Comment (srampal) [inline roles/contiv/templates/contiv.cfg.j2]: One of the current suggested solutions for master HA is to use keepalived, ucarp or equivalent solution to move the first master'sapi server IP address to a second master if the first master goes down. This allows the solution to be independent of needing to access the cluster hostname through skydns (which itself may not yet be up during initial stages of contiv networking bringup).

Comment (sdodson) [inline roles/contiv/templates/contiv.cfg.j2]: openshift.master.cluster_hostname should point at a load balancer, whether that's an ELB, F5, VIP, or other load balancer in front of the API servers. Where as pointing at the ip address of one of the masters would never be load balanced.

**Comment by srampal** [comment]
@sdodson 

If you plan to try this out just to see if the ansible runs successfully and a cluster comes up, go ahead. However if you also want to test actual container data plane traffic, note that vxlan/ ovelay networking mode does not work for the combination of openshift + contiv currently and only vlan encap mode is supported. However this also requires you to setup external physical vlan networking in order for contiv + openshift to work correctly.  We have demoed this to Rajat and Joe from Red Hat to show the full contiv control plane + data plane working correctly with vlan encap mode (and I can show it to you as well if you want to see it). However just a heeads up so that you dont spend time trying to make the data plane work using purely overlay/ vxlan encap mode since that is not currently supported in this particular initial release of openshift + contiv.

**Comment by sdodson** [comment]
Thanks, I'd just be looking to verify that ansible runs without error.

File: roles/contiv/tasks/ovs.yml (hunk 13 above)
Comment (srampal) [inline roles/contiv/tasks/ovs.yml]: I believe this is stored in the ovs configuration database persistently.

File: roles/contiv/templates/contiv.cfg.j2 (hunk 15 above)
Comment (srampal) [inline roles/contiv/templates/contiv.cfg.j2]: Agreed. I tested earlier by using "openshift.default.svc.cluster.local" and other such options in the above j2 template but those didnt seem to work (possibly because contiv needs to be able to access the kubernetes api right in the beginning when even DNS is not reachable or has not populated these domain name entries into the local DNS yet). Hence I reverted to using the IP address with the expectation of a VIP based solution for HA (keepalived/ ucarp/ pacemaker etc). Agreed that this limits our ability to load balance the API server for now but at least we know it works reliably in our testing. Could we visit the load balancing after this merge so that we determine a solution that works reliably in our testing and at that time update this j2 template if needed ? In any case someone can always over-ride this variable through the inventory file if they have a scenario that works reliably with using a cluster DNS name rather than IP address.

File: roles/contiv/handlers/main.yml (hunk 4 above)
Comment (srampal) [inline roles/contiv/handlers/main.yml]: @DanyC97 

I agree in principle but even the different roles within Openshift itself are not consistent with each other. Compare openshift_cli with openshift_loadbalancer or many other such cases which are different from each other in style/ formatting.

Note that contiv also needs to maintain its own style consistency between what we deliver for Openshift versus other container platforms.


I will make the changes to netmaster.yml and netplugin.yml to be consistent within themselves but we can keep going forever if we want to match with all the different Openshift files as well as all vendor submitted files to have the same formatting style.

File: roles/contiv/tasks/netmaster.yml (hunk 16 above)
Comment (srampal) [inline roles/contiv/tasks/netmaster.yml]: see comment below

**Comment by srampal** [comment]
@sdodson @DanyC97  PTAL   I've made all the suggested changes (plus some more) although a couple of the suggestions had to be backed out because the CI tests were not accepting some of the formatting changes related to indentation.

File: roles/contiv/tasks/pkgMgrInstallers/centos-install.yml (hunk 14 above)
Comment (rajatchopra) [inline roles/contiv/tasks/pkgMgrInstallers/centos-install.yml]: For the 'origin' work, the centos package is fine. As we move to support for OSE, this has to go out. Also applies to enforcing selinux.

File: roles/contiv/templates/netplugin.j2 [hunk 18]
```diff
@@ -0,0 +1,6 @@
+{% if contiv_encap_mode == "vlan" %}
+NETPLUGIN_ARGS='-vlan-if {{ netplugin_interface }} -ctrl-ip {{ netplugin_ctrl_ip }} -plugin-mode kubernetes -cluster-store etcd://{{ etcd_url }}'
+{% endif %}
+{% if contiv_encap_mode == "vxlan" %}
```
Comment (rajatchopra) [inline roles/contiv/templates/netplugin.j2]: @srampal Do we want to disable this until we get this working?

**Comment by srampal** [comment]
@Rajatchopra

We are adding more detailed documentation in the openshift-docs repo. I can add a read here as well but where would you like me to add it ? Adding at roles/contiv/README.md may not be appropriate since that is for documenting the role and not the overall solution. However I can add something brief there if everyone is okay with it. In any case the solution is quite complex and it will not be possible to include all the details in the ansible repo and the user will need to go to the openshift docs web site as well as contiv web site for full details. Let me know if it is useful to add documentation here versus in the docs repo/ web site.

Will squash commits …

Kube-proxy is not disabled by contiv. Although contiv takes over most of the east –west load balancing functions of kibe-proxy, there are some scenarios that will use the kube-proxy path even when contiv is enabled.

**Comment by srampal** [comment]
@Rajatchopra

I can add a section tomorrow itself similar to this one for contiv into the docs repo. That might be more useful than adding some limited documentation in the ansible repo which will not be very useful to someone wanting to run contiv+ openshift.

https://docs.openshift.org/latest/install_config/configuring_sdn.html

Thoughts ? If that makes sense, lets merge the ansible PR and focus on docs separately for the docs repo.


Rgds,
Sanjeev


From: srampal <srampal@cisco.com<mailto:srampal@cisco.com>>
Date: Wednesday, February 22, 2017 at 7:42 PM
To: openshift/openshift-ansible <reply@reply.github.com<mailto:reply@reply.github.com>>, openshift/openshift-ansible <openshift-ansible@noreply.github.com<mailto:openshift-ansible@noreply.github.com>>
Cc: Mention <mention@noreply.github.com<mailto:mention@noreply.github.com>>
Subject: Re: [openshift/openshift-ansible] Pull request for Contiv Ansible code integration into Openshift Ansible (#3393)

@Rajatchopra

We are adding more detailed documentation in the openshift-docs repo. I can add a read here as well but where would you like me to add it ? Adding at roles/contiv/README.md may not be appropriate since that is for documenting the role and not the overall solution. However I can add something brief there if everyone is okay with it. In any case the solution is quite complex and it will not be possible to include all the details in the ansible repo and the user will need to go to the openshift docs web site as well as contiv web site for full details. Let me know if it is useful to add documentation here versus in the docs repo/ web site.

Will squash commits …

Kube-proxy is not disabled by contiv. Although contiv takes over most of the east –west load balancing functions of kibe-proxy, there are some scenarios that will use the kube-proxy path even when contiv is enabled.

**Comment by DanyC97** [comment]
@srampal LGTM for now, there are still items under `with_item` block which are not all aligned in the same way however get this in and i can push a PR after that to get the consistency across.

**Comment by jcpowermac** [comment]
LGTM

File: playbooks/adhoc/contiv/delete_contiv.yml (hunk 3 above)
Comment (srampal) [inline playbooks/adhoc/contiv/delete_contiv.yml]: I made this change earlier but the Travis CI tests were failing with the new indentation so had to back out your suggestion and went back to the original indentation which did pass the CI tests.

File: roles/contiv/tasks/netmaster_iptables.yml (hunk 8 above)
Comment (srampal) [inline roles/contiv/tasks/netmaster_iptables.yml]: Made the changes and then backed out (because of Travis CI unit tests failing) as noted on prior comment.

File: roles/contiv/tasks/netmaster.yml (hunk 17 above)
Comment (srampal) [inline roles/contiv/tasks/netmaster.yml]: Done

File: roles/contiv/templates/netplugin.j2 (hunk 18 above)
Comment (srampal) [inline roles/contiv/templates/netplugin.j2]: @rajatchopra 

1) Added a README file under roles/contiv.  Anyway still need proper documentation in the Openshift docs repo which we will add in the coming days/ weeks.

2) Briefly mentioned in this README but referred to Openshift docs for full details

3) Squashed

4) We retain kube-proxy for some Openshift scenarios

5) Not straightforward to disable VXLAN mode specially since it can be used in some limited scenarios specially with native kubernetes over cloud. Documentation clearly states that only VLAN mode and bare metal deployment is supported currently for Openshift + Contiv.  Will remove these restrictions in coming weeks.

**Comment by srampal** [comment]
@DanyC97 

As mentioned in above notes, the suggested indentation of with_items had been made earlier in response to your comments. But that change was causing Travis CI unit tests to fail. Hence I had to back out the changes to get the tests to pass. Other changes (native yaml formatting etc) have been made.

File: roles/contiv/tasks/delete_contiv.yml (hunk 1 above)
Comment (srampal) [inline roles/contiv/tasks/delete_contiv.yml]: moved (and replied to this) earlier ... just adding reply here now also directly under the original comment for completeness.

File: roles/contiv/tasks/cleanup.yml (hunk 2 above)
Comment (srampal) [inline roles/contiv/tasks/cleanup.yml]: Removed roles/contiv/cleanup.yml. Use playbooks/adhoc/contiv/delete_contiv.yml

**Comment by rajatchopra** [comment]
@srampal ack on README, kube-proxy!
If you think you can quickly disable the vxlan mode (because it does not work right now), I would prefer that.
LGTM, nevertheless

**Comment by sdodson** [comment]
aos-ci-test

**Comment by DanyC97** [comment]
anything holding this Pr @sdodson ? have you found any issues on your local tests?

**Comment by sdodson** [comment]
Going ahead with the merge.

**Comment by DanyC97** [comment]
@sdodson any insights as to which openshift-ansible release/ tag this will PR will make it?

**Comment by zhwenh** [comment]
openshift_use_openshift_sdn=false
openshift_use_contiv=true
os_sdn_network_plugin_name='cni'

I get the last version and setup with ansible show this error:

TASK [lib_openshift : lib_openshift detect ostree] *****************************
fatal: [10.1.130.14]: FAILED! => {
    "failed": true
}

MSG:

The conditional check 'contiv_role == "netmaster"' failed. The error was: error while evaluating conditional (contiv_role == "netmaster"): 'contiv_role' is undefined

**Comment by DanyC97** [comment]
hi @zhwenh 

sorry for the issue, can you pls confirm your last openshift-ansible` git repo (master branch) does contain this [PR](https://github.com/openshift/openshift-ansible/pull/3546)

**Comment by srampal** [comment]
Hi @zhwenh yes, I noticed that and am investigating. Maybe something broke in the contiv installer during the merge ... it was working correctly on the last PR version of the code.

**Comment by sdodson** [comment]
@DanyC97 That PR should only affect the installer when it's been built into RPMs and installed that way which I doubt is the case but maybe I've missed something.

**Comment by zhwenh** [comment]
**I get the release  openshift-ansible-3.5.26-1**
openshift_use_openshift_sdn=false
openshift_use_contiv=true
os_sdn_network_plugin_name='cni'

setup with ansible show this error:

TASK [contiv : Netplugin IPtables | Enable iptables at boot] *******************
fatal: [10.1.130.15]: FAILED! => {
    "changed": false, 
    "failed": true
}

MSG:

Unable to enable service iptables: Failed to execute operation: Cannot send after transport endpoint shutdown


fatal: [10.1.130.14]: FAILED! => {
    "changed": false, 
    "failed": true
}

**Comment by srampal** [comment]
Can you please try with setting the following in your inventory file ?

os_firewall_use_firewalld=false


Rgds,
Sanjeev

**Comment by srampal** [comment]
@zhwenh  In addition to setting 

os_firewall_use_firewalld=false 

I recommend you also use the fix from the PR 

https://github.com/openshift/openshift-ansible/pull/3586#pullrequestreview-25613788

which we need for contiv to work properly.
